source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -482,11 +482,193 @@ account:
         let amount1 = Amount::from_ccd(20);
         claim_eq!(host.get_transfers_to(receiver0), [amount0, amount1]);
 
+.. _writing_property_based_tests:
+
+Writing property-based tests
+============================
+
+The property-based testing technique allows for testing statements about your code that are expected to be true for any input parameters.
+The input to such tests is generated randomly.
+Property-based testing is supported using the |QuickCheck|_ crate.
+The tests should be placed in the same module as regular unit tests and annotated with the ``#[concordium_quickcheck]`` macro.
+The return value of the function should be a boolean corresponding to whether the property holds.
+
+To get started, add the ``concordium-quickcheck`` feature to ``concordium-std`` as a ``dev``-dependency in ``Cargo.toml``:
+
+.. code-block::
+
+    ...
+
+    [dev-dependencies]
+    concordium-std = { version = "5", features = ["concordium-quickcheck"] }
+
+    ...
+
+The ``concordium_quickcheck`` macro takes the ``num_tests`` attribute for specifying the number of random tests to run.
+In the code snippet below, the parameters ``address`` and ``amount`` are generated randomly.
+The process of generating random input and running the test is repeated ``num_tests = 500``.
+
+.. code-block:: rust
+
+    #[concordium_cfg_test]
+    mod test {
+
+       #[concordium_quickcheck(num_tests = 500)]
+       fn some_property_test(address: Address, amount: Amount) -> bool {
+        ...
+        // Instantiate custom struct with random parameters, if necessary.
+        let input = MyParameters { sender: address, payment: amount }
+        ...
+        }
+    }
+
+The types ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
+Read more about available ``Arbitrary`` instances for Concordium-specific types in |concordium_contracts_common|_ documentation.
+|QuickCheck|_ defines ``Arbitrary`` instances for standard data types, like numbers and collections (``Vec``, ``BTreeMap``, etc.).
+These instances are available by default when writing tests.
+Custom user data type instances, like ``MyParameters`` above, can be created directly in tests using the random input parameters or by defining ``Arbitrary`` instances.
+See more details on QuickCheck's ``Arbitrary`` `here <https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html>`_.
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    $cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
```

Comment:
Please explain here that all the random inputs are generated deterministically from the random seed. That is implied here, but it would be good to make it explicit for non-experts.

Reply:
Done.